### DISCUSSION TITLE
(RESOLVIDO) Como rodar o Python no cPanel de Hospegadem compartilhada?

### DISCUSSION THREAD

**Discussion by isomiluizdasilva** [andreroggeri/pynubank#328] [General]
Inicialmente já me desculpo, pois sou iniciante nesse mundo Python. Mas como infelizmente não temo a API do banco em PHP que é a linguagem que domino, e estou precisando urgentemente integrar minha conta, baixei o pacote aqui do github, coloque na minha pasta /public_html/_nubank_/, criei o aplicativo Python no cPanel e coloquei para iniciar o passenger_wsgi.py, que nessa situação estaria em /public_html/_nubank_/passenger_wsgi.py, que por sua vez, pelo que entendi, chamaria /public_html/_nubank_/pynubank/nubank.py. Devo ter feito algo errado, pois quando inicio o aplicativo recebo a mensagem:

---
We're sorry, but something went wrong.
The issue has been logged for investigation. Please try again later.

[Technical details for the administrator of this website](http://betget.com.br/_nubank_#)
Error ID: cb7ccf2b
Details:
Web application could not be started by the Phusion Passenger(R) application server.

Please read [the Passenger log file](https://www.phusionpassenger.com/library/admin/log_file/) (search for the Error ID) to find the details of the error.

You can also get a detailed report to appear directly on this page, but for security reasons it is only provided if Phusion Passenger(R) is run with environment set to development and/or with the friendly error pages option set to on.

For more information about configuring environment and friendly error pages, see:
---

Como devo proceder para executar nesse meu servidor de hospedagem compartilhado via cPanel?
Desde já, parabenizo pela iniciativa e agradeço o auxílio.
![capturatela_04072022_022341](https://user-images.githubusercontent.com/61293748/177086732-19aa0975-0358-4e5c-bbc0-a85d985d7e94.png)

**Comment by andreroggeri** [comment]
Você chegou a dar uma olhada na implementação em [PHP](https://github.com/nexy/phpnubank) ? Talvez seja mais fácil do que dessa forma .

De qualquer maneira, pra conseguir ajudar série necessário ver os logs de erro

**Reply by isomiluizdasilva** [reply]
Ótimo, obrigado. não conhecia a versão via PHO, irei verificar, quanto ao log, segue:
---
App 2450126 output: Traceback (most recent call last):
App 2450126 output:   File "/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/wsgi-loader.py", line 369, in <module>
App 2450126 output:     app_module = load_app()
App 2450126 output:   File "/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/wsgi-loader.py", line 76, in load_app
App 2450126 output:     return imp.load_source('passenger_wsgi', startup_file)
App 2450126 output:   File "/home/betgetco/public_html/_nubank_/passenger_wsgi.py", line 8, in <module>
App 2450126 output:     wsgi = imp.load_source('wsgi', 'passenger_wsgi.py')
App 2450126 output:   File "passenger_wsgi.py", line 8, in <module>
App 2450126 output:     wsgi = imp.load_source('wsgi', 'passenger_wsgi.py')
App 2450126 output:   File "passenger_wsgi.py", line 8, in <module>
App 2450126 output:     wsgi = imp.load_source('wsgi', 'passenger_wsgi.py')
App 2566779 output:   File "passenger_wsgi.py", line 8, in <module>
App 2566779 output:     wsgi = imp.load_
App 2566779 output: source('wsgi', 'passenger_wsgi.py')
App 2566779 output:   File "passenger_wsgi.py", line 8, in <module>
App 2566779 output:     wsgi = imp.load_source('wsgi', 'passenger_wsgi.p
App 2566779 output: y')
App 2566779 output:   File "passenger_wsgi.py", line 1, in <module>
App 2566779 output:     import imp
App 2566779 output: RuntimeError: maximum recursion depth exceeded while calling a Python object
---
[logo2.log](https://github.com/andreroggeri/pynubank/files/9039309/logo2.log)

**Reply by andreroggeri** [reply]
Não vejo nenhum log referente ao `pynubank`. Pelo o que me parece seu arquivo `passenger_wsgi.py` está importando ele mesmo e entrando em recursao infinita

**Reply by isomiluizdasilva** [reply]
Valeu André, Muito obrigado pela rapidez e eficiências nas respostas. Consegui com seu auxílio fazer rodar a versão em PHP, estou tentando ver as semelhanças com esse projeto aqui para ver se descubro como detalhar o extrato, pois está aparecendo somente o valor do pix recebido e a data, preciso ver o ID e o identificador enviado. Se puder me auxiliar, novamente lhe agradeço. É que lá no outro projeto phpnubank não vi a área de fórum

**Reply by andreroggeri** [reply]
Eu não manjo de PHP, você pode usar a seção de issues pra tirar essa dúvida com o autor da implementação.

**Reply by isomiluizdasilva** [reply]
Valeu!!!! Saberia então me dizer onde vejo uma documentação, que mencione essa variáveis disponibilizadas pela NUBANK e que são utilizadas tanto em seu projeto PY quanto em PHP:
Exemplo:
savingsAccount / currentSavingsBalance / netAmount
ou
['savingsAccount']['currentSavingsBalance']['netAmount']

**Comment by andreroggeri** [comment]
Infelizmente essa api do nubank não é documentada.

Tudo que foi feito aqui foi pegando inspecionando os requests que o app faz
em cada tela

Em seg., 4 de jul. de 2022 12:25, Isomi Luiz da Silva <
***@***.***> escreveu:

**Reply by isomiluizdasilva** [reply]
A falta de documentação dificulta um pouco. Mas na sua API em Phyton é listado extrato com as informação de Identificador/Descrição e e ID da Transação, como mostrado nos comprovantes do App. O que preciso na realidade é identificar o PIX recebido, já consigo ter acesso a conta via API em PHPNUBANK, o pix gerado e enviado é identificado, como mostrado no comprovante em anexo. mas na api ao acessar, mostra somente DATA e VALOR, preciso mostrar a DESCRIÇÃO ou o ID DA TRANSAÇÃO. Conseguiria esse parâmetro na API em Phyton?
Conseguindo o parâmetro tento ver se consigo aqui em PHP, mas se na sua já conseguir ter acesso a essas informações, terei de me acabar por aqui para entender Phyton ao menos para fazer essa API funcionar :) :) :) :)

**Comment by andreroggeri** [comment]
Isso dá pra fazer, mas na lista de transações o identificador do pix não é
retornado. Pra isso você tem que fazer uma chamada com uma query específica
que vai trazer os detalhes da transação (e aí vai ter um campo com o
identificador pix).

Mas acaba sendo inviável pq você vai ter que fazer uma chamada de detalhe
pra cada transação, o que provavelmente vai fazer você bloquear a sua conta
dependendo da quantidade .

Em seg., 4 de jul. de 2022 12:41, Isomi Luiz da Silva <
***@***.***> escreveu:

**Reply by isomiluizdasilva** [reply]
Não sou expert nessa área de banco :) :). Acho que já deu para notar :) :)
Será que teríamos uma dessas opções:
1. Exportar todos os recibos por período
2. Gerar o QRCode que ao ser pago, nos retorne o ID da transação?

Hoje eu gero o QRCode com identificado e descrição que é enviado, mas ainda não consegui pegar o retorno no ID da transação no ato do pagamento. Isso também resolveria, assim nem precisaria acessar os dados da conta para o que preciso que seria pegar o ID ou IDENTIFICADOR.

Em resumo preciso é confirmar o recebimento do PIX. Daí a saga :)

**Reply by andreroggeri** [reply]
[aqui](https://github.com/andreroggeri/pynubank/blob/master/examples/pix.md) tem um exemplo de como recuperar identificador pix uma transação (se ela for pix e tiver o identificador)

**Reply by isomiluizdasilva** [reply]
Pelo que entendi o código Phyton, faz um loop e pegar 100 últimos transaction['id'], certo?  Ou ele pega o ID de uma única operação?
---
from pynubank import Nubank, MockHttpClient

nu = Nubank(MockHttpClient())
nu.authenticate_with_qr_code('some-cpf', 'some-password', 'some-uuid')

transactions = nu.get_account_statements()

for transaction in transactions[:100]:
    tx_status = nu.get_pix_identifier(transaction['id'])
    # Poderá retornar None caso a transação não seja Pix ou não tiver identificador
    # Caso contrário retorna o identificador único que foi cadastrado no QRCode
    print(tx_status) 
---

**Reply by andreroggeri** [reply]
Isso, das últimas 100, mas poderia ser de todas se não tivesse limitando (mas provavelmente a API iria bloquear a sua conta se fizer tantas requisições assim)

**Reply by isomiluizdasilva** [reply]
Já, acabei de testar, em PHH também funciona, só que como está na ta
![capturatela_04072022_133834](https://user-images.githubusercontent.com/61293748/177193598-56d5db03-623d-4cc6-a6e2-229062e8e64f.png)
bela de feed, pega o ID de index do feed e não do extrato, dessa forma, voltamos para aquela situação em que será necessário gerarmos uma chamada para cada ID do feed para só então termos acesso ao ID da Transação.

**Reply by isomiluizdasilva** [reply]
Esses são os dados possíveis via feed:
Array ( [0] => id [1] => __typename [2] => title [3] => detail [4] => postDate [5] => amount [6] => originAccount )

![capturatela_04072022_140007](https://user-images.githubusercontent.com/61293748/177195900-f01602a3-142d-4d1f-bc3b-c77827665c42.png)

Mas chegaremos lá, será que realmente se eu fizer um loop por cada um dos IDs do feed pode ser bloqueado a conta :) :) :)    ?
Os feeds eu procuraria em ['viewer']['savingsAccount']['feed'] e os comprovantes do PIX pixTransaction

**Reply by andreroggeri** [reply]
É, essa transação que você mandou provavelmente não é pix, e sim doc/Ted.

Pix o typename é GenericFeedEvent

Sobre as limitações da API, teria que testar, eu não consigo te dar garantia nenhuma, mas nessa issue compartilharam algumas experiências https://github.com/andreroggeri/pynubank/issues/238

**Reply by isomiluizdasilva** [reply]
Todos foram PIX, fui eu quem fiz via leitura do QRCode gerado na página do Nubank! Vou continuar fossando aqui. Vou ter que sair para minha atividade do dia a dia, após 18h, voltar para saga. O dia vai ser pequeno :) :) :)
Valeu mesmo!!!!

**Reply by isomiluizdasilva** [reply]
Primeira etapa concluída:
![capturatela_04072022_191744](https://user-images.githubusercontent.com/61293748/177221600-bd9a3d48-cf5e-489f-89e2-1ee355dd326c.png)
Agora é só ver como rodar!!! :) :)

**Reply by isomiluizdasilva** [reply]
Já consegui rodar a gerar os arquivos .json com todos os feeds, agora é trabalhar apenas no feed desejado, que são as transferências recebidas via PIX.

Em Phyton:
Como eu Faria para rodar o script em intervalos de 10 minutos
---
import json
from pynubank import Nubank, MockHttpClient

nu = Nubank(MockHttpClient())
nu.authenticate_with_cert('xxx', 'xxxx', 'xxxxx.p12')

has_next_page = True
current_page_number = 1
cursor = None

# A lógica abaixo irá recuperar todas as páginas
# E salvar em arquivos sequenciais como feed_1.json, feed_2.json, etc
while has_next_page:
    feed = nu.get_account_feed_paginated(cursor)

    with open(f'feed_{current_page_number}.json', 'w') as f:
        json.dump(feed, f)

    has_next_page = feed['pageInfo']['hasNextPage']
    cursor = feed['edges'][-1]['cursor']
    current_page_number += 1
---

Exemplo que utilizei para gerar mensagem:
--
import sched
scheduler = sched.scheduler()

def printa():
    print("Estou rodando no servidor!!")
    scheduler.enter(delay=5, priority=0, action=printa)
    
printa()

scheduler.run(blocking=True)
--
